packages/graalvm-java/src/main/kotlin/elide/runtime/plugins/java/shell/GuestJavaEvaluator.kt

```diff
@@ -0,0 +1,89 @@
+import org.graalvm.polyglot.Source
+import kotlin.streams.asSequence
+import elide.runtime.core.DelicateElideApi
+import elide.runtime.core.GuestLanguageEvaluator
+import elide.runtime.core.PolyglotContext
+import elide.runtime.core.PolyglotValue
+
+/**
+ * An interactive Java interpreter capable of compiling and evaluating snippets of code, for use in a REPL. Use the
+ * [evaluate] method to execute a Java snippet and receive the result as a [PolyglotValue].
+ *
+ * This class uses [JShell] with a [GuestExecutionControl] backend, which wraps around a guest [context] to execute
+ * code using Espresso. Some current limitations apply:
+ * - Startup is currently very slow: both because of Espresso's experimental state *and* JShell's normal overhead.
+ * - Return values are not supported yet, since JShell only returns [String] as an evaluation result.
+ *
+ * Note that while a [GuestExecutionProvider] is implemented as part of the internal pipeline, it cannot be used with
+ * JShell's service loader, since it requires a [PolyglotContext] as a constructor argument.
+ *
+ * #### How it works
+ *
+ * JShell separates snippet compilation from execution, allowing custom execution backends to be implemented. The
+ * default implementation simply runs the snippet in the same JVM, while others may choose to send snippets over the
+ * network or to a different process for execution.
+ *
+ * In this case, the [GuestExecutionControl] implementation wraps around a `LocalExecutionControl` instance created in
+ * a guest context, and delegates all method calls to that instance. This means snippets are executed in the Espresso
+ * context rather than the host.
+ *
+ * @see GuestExecutionProvider
+ * @see GuestExecutionControl
+ */
+@DelicateElideApi internal class GuestJavaEvaluator(private val context: PolyglotContext) : GuestLanguageEvaluator {
+  /** A cached [JShell] instance configured with the [GuestExecutionProvider], using the [context] for execution. */
+  private val shell: JShell by lazy {
+    JShell.builder().executionEngine(
+      /* executionControlProvider = */ GuestExecutionProvider(context),
+      /* executionControlParameters = */ mapOf(),
+    ).build()
+  }
+
+  /**
+   * Attempt to "interpret" this [event], throwing an exception if it represents an error, or returning a value if it
+   * provides one.
+   */
+  private fun interpret(event: SnippetEvent): String? = when (event.status()) {
+    DROPPED, REJECTED -> {
+      val diagnostics = shell.diagnostics(event.snippet()).asSequence().joinToString("\n") { it.getMessage(null) }
+      error("Snippet evaluation resulted in error: $diagnostics")
+    }
+
+    else -> {
+      event.exception()?.let { throw it }
+      event.value()
+    }
+  }
+
+  override fun accepts(source: Source): Boolean {
+    // binary sources (e.g. jvm bytecode) are not allowed, and only interactive (shell) mode is supported
+    return source.isInteractive && source.hasCharacters()
+  }
```

Comment:
The custom evaluator for the `Java` plugin is a special case: since the language id used by GraalVM for Espresso is `java`, the `Jvm` and `Java` plugins share the same language id. To avoid interference in evaluation between these plugins in the future, sources are only evaluated as Java snippets if they are marked as `interactive` *and* they are character-based.

Compiled JVM bytecode will usually be provided binary-backed sources, which will be rejected by the custom evaluator, leaving them for Espresso to handle. Note that Espresso does not currently support that use case either (arbitrary bytecode execution), but if/when it becomes available, we will be able to avoid clashes.

Reply:
So, in that case, is `Jvm` for bytecode evaluation, and `Java` et al for source evaluation for a given language? Or, what's the distinction between the two engines, and why would we keep both?

Reply:
Yes, `Jvm` merely configures and enables Espresso for a context. If Espresso supports bytecode evaluation in the future, that's the entry point (same as other base language plugins like JS). Until then, binary sources with language id `java` will not be evaluated, while character-based sources are handled by the `Java` plugin's shell.

 I've considered merging both plugins since they seem to just complement each other, what do you think?.